drivers/flash/flash_hp_ra.c

```diff
@@ -0,0 +1,390 @@
+{
+	struct flash_hp_ra_data *flash_data = dev->data;
+
+	if (!flash_ra_valid_range(flash_data->area_size, offset, len)) {
+		return -EINVAL;
+	}
+
+	if (!len) {
+		return 0;
+	}
+
+	LOG_DBG("flash: read 0x%lx, len: %u", (long)(offset + flash_data->area_address), len);
+
+	memcpy(data, (uint8_t *)(offset + flash_data->area_address), len);
+
+	return 0;
+}
+
+static int flash_ra_erase(const struct device *dev, off_t offset, size_t len)
+{
+	struct flash_hp_ra_data *flash_data = dev->data;
+	struct flash_hp_ra_controller *dev_ctrl = flash_data->controller;
+	static struct flash_pages_info page_info_off, page_info_len;
+	fsp_err_t err = FSP_ERR_ASSERTION;
+	uint32_t block_num;
+	int rc, rc2;
+
+	if (!flash_ra_valid_range(flash_data->area_size, offset, len)) {
+		return -EINVAL;
+	}
+
+	if (!len) {
+		return 0;
+	}
+
+	LOG_DBG("flash: erase 0x%lx, len: %u", (long)(offset + flash_data->area_address), len);
+
+	rc = flash_get_page_info_by_offs(dev, offset, &page_info_off);
+
+	if (rc != 0) {
+		return -EIO;
+	}
+
+	if (offset != page_info_off.start_offset) {
+		return -EIO;
+	}
+
+	rc2 = flash_get_page_info_by_offs(dev, (offset + len), &page_info_len);
+	if (rc2 != 0) {
+		return -EIO;
+	}
+
+#if defined(CONFIG_DUAL_BANK_MODE)
+	/* Invalid offset in dual bank mode, this is reversed area. */
+	if ((page_info_off.index > FLASH_HP_CF_BLOCK_32KB_DUAL_LOW_END &&
+	     page_info_off.index < FLASH_HP_CF_BLOCK_8KB_HIGH_START) ||
+	    (page_info_len.index > FLASH_HP_CF_BLOCK_32KB_DUAL_LOW_END &&
+	     page_info_len.index < FLASH_HP_CF_BLOCK_8KB_HIGH_START)) {
+		return -EIO;
+	}
+#endif
+
+	if ((offset + len) != (page_info_len.start_offset)) {
+		return -EIO;
+	}
+
+	block_num = (uint32_t)((page_info_len.index) - page_info_off.index);
+
+	if (block_num > 0) {
+		k_sem_take(&dev_ctrl->ctrl_sem, K_FOREVER);
+
+		err = R_FLASH_HP_Erase(&dev_ctrl->flash_ctrl,
+				       (long)(flash_data->area_address + offset), block_num);
+
+		if (flash_data->FlashRegion == DATA_FLASH) {
+			/* Wait for the erase complete event flag, if BGO is SET  */
+			if (true == dev_ctrl->fsp_config.data_flash_bgo) {
+				while (!g_event_flash.erase_complete)
+					;
+				g_event_flash.erase_complete = false;
+			}
+		}
+
+		k_sem_give(&dev_ctrl->ctrl_sem);
+	}
+
+	if (err != FSP_SUCCESS) {
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static int flash_ra_write(const struct device *dev, off_t offset, const void *data, size_t len)
+{
+	fsp_err_t err = FSP_ERR_ASSERTION;
+	struct flash_hp_ra_data *flash_data = dev->data;
+	struct flash_hp_ra_controller *dev_ctrl = flash_data->controller;
+
+	if (!flash_ra_valid_range(flash_data->area_size, offset, len)) {
+		return -EINVAL;
+	}
+
+	if (!len) {
+		return 0;
+	}
+
+	LOG_DBG("flash: write 0x%lx, len: %u", (long)(offset + flash_data->area_address), len);
+
+	k_sem_take(&dev_ctrl->ctrl_sem, K_FOREVER);
+
+	err = R_FLASH_HP_Write(&dev_ctrl->flash_ctrl, (uint32_t)data,
```

Comment:
Same question regarding the err as in erase.

Reply:
Same as above. I updated. Thanks.